File: blockchain_state.go

package main

import (
    "crypto/sha256"
    "encoding/hex"
    "fmt"
    "strconv"
    "time"
)

var Blockchain []Block

// Estrutura de um bloco
type Block struct {
    Index        int
    Timestamp    string
    Data         string
    PreviousHash string
    Hash         string
}

// Função para calcular o hash de um bloco
func calculateHash(block Block) string {
    record := strconv.Itoa(block.Index) + block.Timestamp + block.Data + block.PreviousHash
    hash := sha256.New()
    hash.Write([]byte(record))
    hashed := hash.Sum(nil)
    return hex.EncodeToString(hashed)
}

// Função para criar um novo bloco
func createBlock(prevBlock Block, data string) Block {
    newBlock := Block{
        Index:        prevBlock.Index + 1,
        Timestamp:    time.Now().String(),
        Data:         data,
        PreviousHash: prevBlock.Hash,
        Hash:         "",
    }
    newBlock.Hash = calculateHash(newBlock)
    return newBlock
}

func addBlock(newBlock Block) {
    Blockchain = append(Blockchain, newBlock)
}

func main() {
    // Criando o bloco gênese
    genesisBlock := Block{0, time.Now().String(), "Bloco Gênese", "", ""}
    genesisBlock.Hash = calculateHash(genesisBlock)
    Blockchain = append(Blockchain, genesisBlock)

    // Adicionando novos blocos
    newBlock := createBlock(genesisBlock, "Transação 1: Andrej enviou 10 moedas para BabaYaga")
    addBlock(newBlock)

    fmt.Println("Blockchain:", Blockchain)
}

File: blockchain_state.py

import hashlib
import time

class Block:
    def __init__(self, index, timestamp, data, previous_hash):
        self.index = index
        self.timestamp = timestamp
        self.data = data
        self.previous_hash = previous_hash
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        data_to_hash = f"{self.index}{self.timestamp}{self.data}{self.previous_hash}"
        return hashlib.sha256(data_to_hash.encode()).hexdigest()

def create_block(previous_block, data):
    index = previous_block.index + 1
    timestamp = str(time.time())
    return Block(index, timestamp, data, previous_block.hash)

def add_block_to_chain(blockchain, block):
    blockchain.append(block)

if __name__ == "__main__":
    blockchain = [Block(0, str(time.time()), "Bloco Gênese", "0")]

    new_block = create_block(blockchain[0], "Transação 1: Andrej enviou 10 moedas para BabaYaga")
    add_block_to_chain(blockchain, new_block)

    for block in blockchain:
        print(f"Bloco: {block.__dict__}")
